Criollo/Source/CRConnection.m

```diff
@@ -197,7 +197,10 @@ - (void)sendDataToSocket:(NSData *)data forRequest:(CRRequest *)request {
         return;
     }
     
-    CRRequest* firstRequest = self.requests.firstObject;
+    CRRequest* firstRequest;
```

Comment:
I agree that this is probably what causes the "index 0" exception, but I would like to avoid a lock, they're quite expensive  :).

One option could be copy the `requests` array, and then get the first object. This would ensure that we have a snapshot of it so we don't care if someone messes around with it in the meantime.

`CRRequest* firstRequest = self.requests.copy.firstObject;`

I'm curious to see if that fixes it.

Reply:
I doubt `-[NSArray copy]` will be more thread-safe than `-[NSArray firstObject]`, but I'll try it.

Reply:
As I suspected `copy` is even more fragile.  It throws this one almost immediately ```'NSRangeException', reason: '*** -[__NSArrayM getObjects:range:]: range {0, 1} extends beyond bounds for empty array'```

Reply:
Cool. Thanks for tanking the time to try it out.

Should we move  the assignment to the isolation queue? Sync of course.

Reply:
If both the blocks that call `firstObject` and `removeObject` were on the queue, and both were `dispatch_sync`, I believe that would fix it.  

I'm not sure that's cheaper than an NSLock or other mutex-ing approach that. But any locking or queuing is going to be an order (or two) of magnitude less time-expensive than the disk and network access.

Reply:
Would  the `removeObject` object call also need to be sync?  The queue is serial. Wouldn't it be enough to sync the `firstObject` call. 

The idea of the queue is that all operations dispatched to it happen in the order in which they were dispatched. If we put all operations that manipulate the requests array on that queue, there would be no need for need for locking.

The `removeObject` part could then be dispatched async since there's no need to block the current thread with it.

What do you think?

Reply:
Confirmed. The following pair work.
```
    __block CRRequest* firstRequest;
    dispatch_sync(self.isolationQueue, ^{
        firstRequest = self.requests.firstObject;
    });
.
.
.
- (void)didFinishResponseForRequest:(CRRequest *)request {
    CRConnection * __weak connection = self;
    [self.delegate connection:self didFinishRequest:request response:request.response];
    dispatch_sync(self.isolationQueue, ^{
        [connection.requests removeObject:request];
    });
}

Reply:
Does it still work with the second one async? This is purely out of curiosity.

Reply:
But push it either way: 

- both sync 
- `firstObject` sync and `removeObject` async if it works and/or if you're in the mood to try.